### ISSUE TITLE
When will TensorFlow be supported?

### ISSUE THREAD

**Issue by reiinakano** [onnx/onnx#3]
When will TensorFlow be supported?

**Comment by Yangqing** [comment]
We are looking into TensorFlow support, but there is no set timeline yet. In addition we are definitely interested in working with every software and hardware vendors to enhance this open collaboration.

**Comment by soumith** [comment]
let's keep the issue open. I think we'll track the TensorFlow support in this issue. We'll communicate here if we make any progress.

On that note, if anyone in the community builds TensorFlow support for ONNX, do let us know.

**Comment by binarybana** [comment]
Is the plan to build ONNX connectors in separate repositories under the ONNX github org (eg caffe2-onnx)?

**Comment by soumith** [comment]
we haven't fully decided that. Right now there's onnx/onnx-caffe2, but the pytorch bits are in pytorch/pytorch. We are debating moving onnx-caffe2 out (no clear concensus). what do you think is better?

**Comment by binarybana** [comment]
For neon/ngraph we're probably leaning on putting them in our ngraph repository to be close to the other existing converters and connectors we have.

Not much different either way though besides: one less dependency to pull down for users and easier to make sweeping changes while keeping all the connectors in sync.

**Comment by soumith** [comment]
that sounds good.

**Comment by bhack** [comment]
Is this an alternative for https://github.com/tensorflow/tensorflow/issues/4815?

**Comment by ebarsoum** [comment]
Let's open other issues and not party on the same issue.

The link above talk about NNEF: https://www.khronos.org/nnef, ONNX is more than exchange format, we define a list of common OPs and their semantics in which we expect all frameworks supporting ONNX to implement. Also, the focus isn't inference only.

**Comment by bhack** [comment]
It think that it was released cause the topic is very similar. I hope that the TF team  will be interested more in ONNX that what was in NNEF and OPENVX (also NNEF is not just just an exchange format cause was releated to [Openvx Neural network extension](https://www.khronos.org/registry/OpenVX/extensions/neural_network/html/index.html)

**Comment by botev** [comment]
I think for this effort to be well adopted you need a bit more consideration and much larger conversation in the community about it. My personal opinion is that in an actual standard/interface we should not have things like Elu, Relu, LeakyRelu, ParametricRelu, MagicRelu, Selu, WhatNotRelu... If you really want to put forward this as a standard for portability definitely such fine grained operators are highly redundant. This can be seen from any of the graph based frameworks where this operator does not exist as a Graph Op but as the composition of those. Similarly, follows for reduction with associative operators. The whole idea of a standard I think has been going around for quite a while, but that must be quite general.

In my opinion, such interop is only possible if it goes only on the mathematical level, essentially having a mathematical description of programs which is agnostic to any implementation.

**Comment by Yangqing** [comment]
@botev I respectfully disagree on the fact that we want to go down to the very mathematical description of programs. ONNX is not intended to be a general purpose math expression library and is specifically focused on neural networks, so finding a proper abstraction that does not make optimization unnecessarily hard is going to be preferred. Of course, the exact balance is something needing detailed discussion.

As an example, Dropout can be implemented as

mask = Rand(m, n)
Y = X * mask * scale

but in a neural network sense this probably isn't the best way for us to represent the neural network semantics.

**Comment by botev** [comment]
@Yangqing It's always good to have a discussion, I'm expressing here only my own personal view on this topic. But then at what point do we decide that it will be too granular? Do you plan to have things like cudNNConv vs CorrMMConv. Additionally, if we don't have the option of the mathematical level does this mean that newly published researcher on architectures, which often includes an ad-hoc and very new combination of operations, will essentially be not supported (consider a new recurrent cell like the weighted average for instance)?

**Comment by soumith** [comment]
We've debated this a lot while defining the scope of ONNX. We concluded that a CuDNN API like granularity is suited for ONNX. There are factors pulling from two main sides:
- Side 1: a mathematical IR, so that things are extremely expressive (also requested by the compiler efforts)
- Side 2: vendors (particularly mobile vendors) implementing the operators in the IR efficiently

We looked at what the vendors are comfortable with, as well as what the compiler folks are reasonably okay with. In the end, it ended up being a middle-ground with tradeoffs, and looked somewhat at the CuDNN granularity.

**Comment by iamthebot** [comment]
@soumith now that it's still in its infancy, it's worth examining one key point though (let me know if this belongs in a separate issue): once you have dynamic computation graphs (which are prevalent in NLP) you fundamentally have to mathematically define behavior somehow.

Yeah, ONNX really should have a CuDNN like API where certain common operations are well-standardized (which lets hardware vendors build supporting ASICs or instructions) but it would be unwise to leverage the fact that what we have is essentially a graph of self contained (sometimes stateful) computations. If ONNX also provides the primitives for those computations (presumably arithmetic primitives) compilers can become increasingly smart about how to optimize and parallelize those operations. This lets most use cases use the well-defined (and typically hardware accelerated) operations and advanced used cases leverage custom operators that we can still reason about in the IR.

I strongly believe this approach should be taken as support for more sophisticated graphs and operations comes about. Since it allows future architectures and techniques to leverage ONNX without requiring the format to change really frequently to incorporate new techniques.

**Comment by iamthebot** [comment]
@Yangqing I totally agree that completely mathematically formal IR is not the way to go since it becomes awkward at expressing the narrow scope of neural network computations... but why not instead scope the formality somewhat? Once CNNs and RNNs with static graphs are no longer the "hot" thing and we're doing more with dynamic and stochastic graphs we still have a similar computational problem but the connectivity and computations inside nodes in our graphs change. Allowing the IR to allow sufficiently flexible computation in those cases goes a long way.

**Comment by agibsonccc** [comment]
FWIW, I would highly prefer something more grounded in reality than theoretical. Something neural net specific that works in the real world is a great goal to support. I have to agree with @Yangqing here.

I know when we did keras import, we did it because it was a sane target easy to map in software near 1 to 1. I'm hoping to see something similar here.

**Comment by botev** [comment]
@agibsonccc I do not really see how mathematical operations are theoretical or not grounded in any way compared to something `LogSumExp` for instance. Also, I personally don't agree that modern (specifically research oriented) NN architectures are that narrow and the main issue is that they grow with almost every month. Additionally, a lot of people have already recognized that most frameworks actually are very useful for things outside NNs, such as Kernel methods, physics simulations and even partial large scale solvers. I understand that everyone can have different opinions and reasons/goals of what he would want from the project and as such, it is up to the maintainers to make the decision. However, for defining such primitive I think I would agree with @yaroslavvb 's comment on the tensorflow issue that the field is still changing way too fast at this point in time.

**Comment by agibsonccc** [comment]
The broader it is, the less likely something useful will come out of it in the short term. 
If we are going to get broad adoption from the community, not including us but also mxnet, dynet etc..let's try to pick something that's easy to setup/maintain. A lot of the things you're describing could definitely come in later versions, but if the actual first goal of this is deployment like what their blog post is saying, a lot of the goals you're stating could be picked up in further revisions or possibly visited in another discussion? Ultimately, it's the project maintainers call,I'm offering an opinion based on someone who maintains a framework with a  little bit of the mindshare.

**Comment by bhack** [comment]
I think actually we are really going out of topic about TF. It is interesting but I think that we can move last comments on a new issue to let other people to be involved

**Comment by andreh7** [comment]
@fmassa mentioned work on an ONNX -> TF converter here https://github.com/onnx/onnx/issues/70#issuecomment-334539370

**Comment by bhack** [comment]
Please follow PR https://github.com/tensorflow/tensorflow/pull/14310

**Comment by prasanthpul** [comment]
Closing this old issue since we have https://github.com/onnx/tensorflow-onnx and https://github.com/onnx/onnx-tensorflow
